C-Plus-Plus/cp/InversionCountProblem.cpp

```diff
@@ -0,0 +1,127 @@
+//Given an array, find the inversion count for the array.
+//Inversion count of an array:- It indicates how far the array is from being a sorted array. For a sorted array, it is zero.
+//Sorted Array:- An array in arranged in a particular order (for this question its acsending order)
+/*
+  For example:-
+  Input:- arr[] = {1, 5, 2, 9, 3}
+  Output:- 3
+  Inversions are:- 
+  (5, 2), (5, 3), (9, 3)
+  
+  Input:- arr[] = {1, 2, 3, 4, 5}
+  Ouput:- 0 
+*/
+
+/* 
+  Why we use Merge Sort approach?
+  If we will use simple approach, i.e., traverse through the array, and for every index, find the number of smaller elements on its right side of the array, using a nested loop and sum up the counts for all index in the array and print the sum, it will take O(n^2) Time Complexity.
+  Instead, we will use Merge Sort approach(Algorithm given below) to reduce Time Complexity to O(nlogn).
+*/  
+
+/*
+  Algorithm:- 
+   1. The idea is similar to merge sort, divide the array into two equal or almost equal halves in each step until the base case is reached.
+   2. Create a function merge that counts the number of inversions when two halves of the array are merged, create two indices i and j, i is the index for the first half, and j is an index of the second half. if a[i] is greater than a[j], then there are (mid – i) inversions. because left and right subarrays are sorted, so all the remaining elements in left-subarray (a[i+1], a[i+2] … a[mid]) will be greater than a[j].
+   3. Create a recursive function to divide the array into halves and find the answer by summing the number of inversions is the first half, the number of inversion in the second half and the number of inversions by merging the two.
+   4. The base case of recursion is when there is only one element in the given half.
+   5. Print the answer.
+*/
```

Comment:
Rewrite it as,

```
/*
  Algorithm:- 
   1. The idea is similar to merge sort, divide the array into two equal or almost equal 
       halves in each step until the base case is reached.
   2. Create a function merge that counts the number of inversions when two halves of the
       array are merged, create two indices i and j, i is the index for the first half, 
       and j is an index of the second half. if a[i] is greater than a[j], then there are 
       (mid – i) inversions. because left and right subarrays are sorted, so all the remaining 
      elements in left-subarray (a[i+1], a[i+2] … a[mid]) will be greater than a[j].
   3. Create a recursive function to divide the array into halves and find the answer by summing 
      the number of inversions is the first half, the number of inversion in the second half and 
      the number of inversions by merging the two.
   4. The base case of recursion is when there is only one element in the given half.
   5. Print the answer.
*/
```